lib/iris/tests/unit/loading/test_load_problems.py

```diff
@@ -0,0 +1,163 @@
+# Copyright Iris contributors
+#
+# This file is part of Iris and is released under the BSD license.
+# See LICENSE in the root of the repository for full licensing details.
+"""Unit tests for the :class:`iris.loading.LoadProblems` class."""
+
+from traceback import TracebackException
+
+import pytest
+
+from iris.coords import DimCoord
+from iris.cube import Cube
+from iris.loading import LoadProblems
+from iris.warnings import IrisLoadWarning
+
+
+@pytest.fixture
+def error():
+    return ValueError("Example ValueError")
+
+
+@pytest.fixture
+def stack_trace(error):
+    try:
+        raise error
+    except ValueError as raised_error:
+        return TracebackException.from_exception(raised_error)
+
+
+@pytest.fixture(
+    params=[dict, Cube, DimCoord, None],
+    ids=["loaded_dict", "loaded_Cube", "loaded_DimCoord", "loaded_None"],
+)
+def loaded_object(request):
+    lookup = {
+        dict: {"long_name": "foo"},
+        Cube: Cube([1.0], long_name="foo"),
+        DimCoord: DimCoord([1.0], long_name="foo"),
+        None: None,
+    }
+    return lookup[request.param]
+
+
+@pytest.fixture
+def first_filename():
+    return "test.nc"
+
+
+@pytest.fixture
+def problem_instance(first_filename, loaded_object, stack_trace):
+    return LoadProblems.Problem(
+        filename=first_filename,
+        loaded=loaded_object,
+        stack_trace=stack_trace,
+    )
+
+
+@pytest.fixture
+def load_problems_instance(problem_instance):
+    problem2 = LoadProblems.Problem(
+        filename="test2.nc",
+        loaded=problem_instance.loaded,
+        stack_trace=problem_instance.stack_trace,
+    )
+    result = LoadProblems()
+    result._problems = [problem_instance, problem2]
+    return result
+
+
+def test_problem_str(problem_instance):
+    if isinstance(problem_instance.loaded, (Cube, DimCoord)):
+        expected_loaded = problem_instance.loaded.summary(shorten=True)
+    else:
+        expected_loaded = str(problem_instance.loaded)
+
+    expected = (
+        f'{problem_instance.filename}: "{problem_instance.stack_trace}", '
+        f"{expected_loaded}"
+    )
+    assert str(problem_instance) == expected
+
+
+def test_load_problems_str(load_problems_instance):
+    expected_lines = [
+        f"{repr(load_problems_instance)}:",
+        *[f"  {problem}" for problem in load_problems_instance.problems],
+    ]
+    expected = "\n".join(expected_lines)
+    assert str(load_problems_instance) == expected
+
+
+def test_problems_property(load_problems_instance):
+    assert load_problems_instance.problems == load_problems_instance._problems
+
+
+def test_problems_by_file_property(load_problems_instance):
+    expected = {p.filename: [p] for p in load_problems_instance._problems}
+    assert load_problems_instance.problems_by_file == expected
```

Comment:
This bit looks neat, but it doesn't work if there are >1 entries for a given filepath (which you don't have in your test example).  You get a result with a list of only one result (the last entry encountered) for each file.
The implementing code with its loop + setdefault **_does_** work.

Sort-of-equivalent demonstration:
```
>>> # example seems to work ...
>>> ll = [('a', 1, 'this'), ('b', 2, 'that')]
>>> {x[0]: x for x in ll}
{'a': ('a', 1, 'this'), 'b': ('b', 2, 'that')}
>>> 
>>> # expanded, shows it doesn't ...
>>> ll = [('a', 1, 'this'), ('a', 2, 'other'), ('b', 2, 'that')]
>>> {x[0]: x for x in ll}
{'a': ('a', 2, 'other'), 'b': ('b', 2, 'that')}
```

Reply:
93045b35cba8f78ea738262717f9eada9ac8b18d